apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.Data/SpannerDataReader.cs

```diff
@@ -0,0 +1,349 @@
+        }
+
+        /// <inheritdoc />
+        public override bool Read()
+        {
+            return ReadAsync(CancellationToken.None).Result;
+        }
+
+        /// <inheritdoc />
+        public override Task<bool> ReadAsync(CancellationToken cancellationToken)
+        {
+            return ExecuteHelper.WithErrorTranslationAndProfiling(async () =>
+            {
+                await GetMetadataAsync(cancellationToken).ConfigureAwait(false);
+                _innerList.Clear();
+                //read # values == # fields.
+                var first = await _resultset.Next(cancellationToken).ConfigureAwait(false);
+                if (first == null)
+                    return false;
+                _innerList.Add(first);
+                //we expect to get full rows...
+                for (var i = 0; i < _metadata.RowType.Fields.Count - 1; i++)
+                    _innerList.Add(await _resultset.Next(cancellationToken).ConfigureAwait(false));
+
+                return true;
+            }, "SpannerDataReader.Read");
+        }
+
+        /// <inheritdoc />
+        protected override void Dispose(bool disposing)
+        {
+            Logger.LogPerformanceCounter("SpannerDataReader.ActiveCount",
+                () => Interlocked.Decrement(ref s_readerCount));
+
+            _resultset?.Close();
+            _connectionToClose?.Close();
+            base.Dispose(disposing);
+        }
+
+        private async Task<int> GetFieldIndexAsync(string fieldName, CancellationToken cancellationToken)
+        {
+            GaxPreconditions.CheckNotNullOrEmpty(fieldName, nameof(fieldName));
+            if (_fieldIndex == null)
+            {
+                _fieldIndex = new Dictionary<string, int>();
+                var metadata = await GetMetadataAsync(cancellationToken).ConfigureAwait(false);
+                if (metadata != null)
+                {
+                    var i = 0;
+                    foreach (var field in metadata.RowType.Fields)
+                    {
+                        _fieldIndex[field.Name] = i;
+                        i++;
+                    }
+                }
+            }
+            return _fieldIndex[fieldName];
+        }
+
+        private Task<ResultSetMetadata> GetMetadataAsync(CancellationToken cancellationToken)
+        {
+            return ExecuteHelper.WithErrorTranslationAndProfiling(async ()
+                => _metadata ?? (_metadata = await _resultset.GetMetadataAsync(cancellationToken)
+                       .ConfigureAwait(false)), "SpannerDataReader.GetMetadata");
+        }
+
+        private V1.Type GetSpannerFieldType(int i)
+        {
+            var fieldMetadata = GetMetadataAsync(CancellationToken.None).Result.RowType.Fields[i];
+            return fieldMetadata.Type;
+        }
+
+#if NET45 || NET451
+
+        /// <inheritdoc />
+        public override void Close()
+        {
+            _resultset?.Close();
```

Comment:
good catch.  probably should.  doesn't matter all that much since dispose should be called regardless, but it makes sense.  note that connectiontoclose will often be null -- its an option to close the underlying connection when you are done reading.